### PR TITLE
[SPEX] Add builder

### PR DESCRIPTION
**Pull request by Wimmerer**
Not sure if the copying is done correctly. The makefile relies on being part of SuiteSparse, but it isn't currently set up that way.

### CODE REVIEW

**Comment by Wimmerer** [comment]
@giordano any idea why those dependencies are not found only on Windows?

**Comment by giordano** [comment]
In https://dev.azure.com/JuliaPackaging/Yggdrasil/_build/results?buildId=22238&view=logs&j=da726748-225d-55bd-4026-a890e36e8df0&t=3510b2f7-4741-5032-84a1-25c6f19f1e85&l=207 I don't see `-L${libdir}` (with libdir expanded, of course),which explains the linking error. I didn't look at the build system, I don't know what's different from other systems.

**Comment by Wimmerer** [comment]
```
/opt/x86_64-w64-mingw32/bin/../lib/gcc/x86_64-w64-mingw32/9.1.0/../../../../x86_64-w64-mingw32/bin/ld: /opt/x86_64-w64-mingw32/bin/../x86_64-w64-mingw32/sys-root/lib/../lib/libmingw32.a(lib64_libmingw32_a-crt0_c.o): in function `main':
/workspace/srcdir/mingw-w64-v7.0.0/mingw-w64-crt/crt/crt0_c.c:18: undefined reference to `WinMain'
collect2: error: ld returned 1 exit status
make[2]: *** [Makefile:72: /workspace/srcdir/SPEX/lib/libspexutil.1.1.0.dll] Error 1
```

Hmm

**Comment by giordano** [comment]
https://github.com/JuliaPackaging/Yggdrasil/pull/5425#issuecomment-1242994320

**Comment by Wimmerer** [comment]
Apologies. The invocation is

```
ranlib libspexutil.lib
cc  -L/workspace/destdir/bin -L/workspace/destdir/lib -L/workspace/destdir/lib -L/workspace/srcdir/SPEX/lib SPEX_matrix_div.o spex_create_mpq_array.o SPEX_free.o SPEX_realloc.o SPEX_matrix_mul.o spex_create_mpz_array.o SPEX_cumsum.o SPEX_malloc.o spex_sparse_collapse.o SPEX_calloc.o SPEX_matrix_allocate.o spex_sparse_realloc.o spex_cast_array.o spex_create_mpfr_array.o spex_expand_double_array.o SPEX_gmp.o SPEX_matrix_copy.o SPEX_matrix_check.o spex_cast_matrix.o spex_expand_mpfr_array.o SPEX_initialize.o SPEX_matrix_free.o SPEX_check_solution.o SPEX_finalize.o spex_expand_mpq_array.o SPEX_initialize_expert.o SPEX_matrix_nnz.o SPEX_create_default_options.o SPEX_LU_analysis_free.o SPEX_LU_analyze.o -o /workspace/srcdir/SPEX/lib/libspexutil.1.1.0.dll -lm -lsuitesparseconfig -lamd -lcolamd -lm -lgmp -lmpfr
/opt/x86_64-w64-mingw32/bin/../lib/gcc/x86_64-w64-mingw32/9.1.0/../../../../x86_64-w64-mingw32/bin/ld: /opt/x86_64-w64-mingw32/bin/../x86_64-w64-mingw32/sys-root/lib/../lib/libmingw32.a(lib64_libmingw32_a-crt0_c.o): in function `main':
/workspace/srcdir/mingw-w64-v7.0.0/mingw-w64-crt/crt/crt0_c.c:18: undefined reference to `WinMain'
```
Which is definitely missing `-shared`. 

The suspect line of the makefile I believe is:
```
$(AR_TARGET): $(OBJ)
	$(ARCHIVE) $@ $^
	- $(RANLIB) $(AR_TARGET)
```
I don't believe this was ever tested on Windows, so I'm expecting to patch this Makefile quite a bit.

**Comment by Wimmerer** [comment]
Ah no that's not correct. The Makefile command that is called is:

```
$(INSTALL_LIB)/$(SO_TARGET): $(OBJ)
	@mkdir -p $(INSTALL_LIB)
	@mkdir -p $(INSTALL_INCLUDE)
	@mkdir -p $(INSTALL_DOC)
	$(CC) $(SO_OPTS) $^ -o $@ $(LDLIBS)
	( cd $(INSTALL_LIB) ; ln -sf $(SO_TARGET) $(SO_PLAIN) )
	( cd $(INSTALL_LIB) ; ln -sf $(SO_TARGET) $(SO_MAIN) )
	$(CP) ../Include/SPEX_Util.h $(INSTALL_INCLUDE)
	chmod 644 $(INSTALL_INCLUDE)/SPEX_Util.h
```
I believe there are significant structural issues with the Makefile that I'll bring up with Tim.

**Comment by giordano** [comment]
> The suspect line of the makefile I believe is:

You don't have to suspect, when `make` fails it tells you exactly which target, which line, in which file, it was trying to run :wink:

File: S/SPEX/build_tarballs.jl [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+    make library -j${nproc}
+fi
+cp ${WORKSPACE}/srcdir/SPEX/include/SPEX_Left_LU.h ${includedir}
+cp ${WORKSPACE}/srcdir/SPEX/lib/libspexleftlu* ${libdir}
+install_license ${WORKSPACE}/srcdir/SPEX/SPEX/LICENSE.txt
+"""
+
+# These are the platforms we will build for by default, unless further
+# platforms are passed in on the command line
+platforms = supported_platforms(;experimental=true)
```
Comment (giordano) [inline S/SPEX/build_tarballs.jl:40]:
```suggestion
platforms = supported_platforms()
```

File: S/SPEX/build_tarballs.jl [hunk 2]
```diff
@@ -0,0 +1,57 @@
...
+products = [
+    LibraryProduct("libspexutil", :libspexutil),
+    LibraryProduct("libspexleftlu", :libspexleftlu)
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency("SuiteSparse_jll"),
+    Dependency("GMP_jll"),
+    Dependency("MPFR_jll")
```
Comment (giordano) [inline S/SPEX/build_tarballs.jl:52]: #3024

Comment (Wimmerer) [reply S/SPEX/build_tarballs.jl:52]: What counts as "old" Julia versions? Pre 1.6? 

So should I depend on that specific version in all cases?

Comment (giordano) [reply S/SPEX/build_tarballs.jl:52]: Yes, or more simply use the version in Julia v1.6 (that's what we target anyway): https://github.com/JuliaLang/julia/blob/57cd92c5d89cd3e2b66ac4d632e81c1f06562da4/stdlib/MPFR_jll/Project.toml#L3 and https://github.com/JuliaLang/julia/blob/release-1.6/stdlib/GMP_jll/Project.toml#L3

Comment (Wimmerer) [reply S/SPEX/build_tarballs.jl:52]: Correct versions I think

File: S/SPEX/build_tarballs.jl [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+# The products that we will ensure are always built
+products = [
+    LibraryProduct("libspexutil", :libspexutil),
+    LibraryProduct("libspexleftlu", :libspexleftlu)
+]
+
+# Dependencies that must be installed before this package can be built
+dependencies = [
+    Dependency("SuiteSparse_jll"),
+    Dependency("GMP_jll", v"6.2.1+2"),
```
Comment (giordano) [inline S/SPEX/build_tarballs.jl:51]:
```suggestion
    Dependency("GMP_jll", v"6.2.1"),
```

**Comment by Wimmerer** [comment]
Hmm should I relax the MPFR version then? Those are the versions Julia should use I think, which should make those jll's compatible?

**Comment by Wimmerer** [comment]
@giordano these are not the same version numbers from the Julia repo "fake" jlls. Are they still acceptable?

**Pull request merged**